Migrate `eig` from the TH to Aten (CUDA)

antocuni opened the pull request:
Fixes #24553

antocuni committed:
WIP: remove eig from THCTensorMathMagma.cu and put an empty implementation in aten/src/ATen/native/cuda/BatchLinearAlgebra.cu

antocuni committed:
Implement at::native::eig_cuda for the case eigenvectors==false

antocuni committed:
avoid the two memcpy by directly passing pointers inside eigvals_data as the wr,wi arguments to magamEig

antocuni committed:
turn the working copy into column-ordered before calling magmaEig

antocuni committed:
add support for the case eigenvectors==true

t-vi reviewed (commented):
Thank you for working on this!
There seem to be a handful of small things to do before the full review.

antocuni committed:
implement the _out of eig

antocuni committed:
the comment was wrong, it is possible to reach this case

antocuni committed:
kill spurious whitespace

antocuni committed:
kill the old THC-based implementation of eig

antocuni committed:
git merge origin/master; Declarations.cwrap has been deleted, so instead of editing that I had to manually kill _th_eig and _th_eig_out from aten/src/ATen/cuda/LegacyTHFunctionsCUDA.cpp

antocuni commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1263,6 +1289,93 @@ std::tuple<Tensor, Tensor> _symeig_helper_cuda(const Tensor& self, bool eigenvec
+
+  // XXX antocuni: this has been copied from _symeig_helper, but is it
+  // possible to reach this code path at all? I don't think it's possible to
+  // have a square matrix with self.numel() == 0
```
Thank you, you are right. I removed the comment

antocuni commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1464,6 +1577,9 @@ Tensor _lu_solve_helper_cuda(const Tensor& self, const Tensor& LU_data, const Te
 
+
+
+
```
done

antocuni committed:
return the correct shape in the empty case; this fixes test_pack_empty

antocuni committed:
add a test for torch.eig(..., out=...)

pearu commented on test/test_torch.py:
```diff
@@ -15954,6 +15954,25 @@ def test_eig(self, device, dtype):
+        # check that the out= version computes the same values as the normal one
+        out_evals = torch.full_like(evals, fill_value=math.nan)
+        out_evecs = torch.full_like(evecs, fill_value=math.nan)
+        evals2, evecs2 = torch.eig(t, eigenvectors=True, out=(out_evals, out_evecs))
```
Add a test of inplace usage of `out`:
```
assert evals2.data_ptr() == out_evals.data_ptr()
assert evecs2.data_ptr() == out_evecs.data_ptr()
```
See copy/inplace discussion above.

antocuni replied:
I'm already checking that `evals2 is out_evals`, so those two lines are redundant. What I could maybe do is to save `old_evals_data_ptr = out_evals.data_ptr()` **before** the call and check that `out_evals.data_ptr() == old_evals_data_ptr`, to ensure that we didn't reallocate the buffer unless we had to?

t-vi replied:
I think just checking that the correct result (same as evals/evecs) is stored in `out_evecs` and that the `data_ptr` stays the same would be good. I think `evals2 is out_evals` might not be as good as `data_ptr` because I'm not sure we have a guarantee that a given C++ Tensor is only wrapped by one PyTorch object (and if we do, we don't need to test).

pearu replied:
I agree with @t-vi , using `is` operator between pytorch tensors may give false-negative result.

antocuni replied:
fixed by https://github.com/pytorch/pytorch/pull/44105/commits/005a7a55c1627d8b3ccf8fdb6c3c6b81c52d67d8

pearu reviewed (changes requested):
I have some suggestions for avoiding unnecessary copies, especially between different devices that can be expensive.

antocuni committed:
blindly try to fix ROCm build by avoiding using magma_int_t in the prototype of apply_eig<>

antocuni commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1263,6 +1289,94 @@ std::tuple<Tensor, Tensor> _symeig_helper_cuda(const Tensor& self, bool eigenvec
+  }
+
+  // magmaEig returns column-ordered tensors, turn them into row-ordered
+  out_eigvals.transpose_(0, 1);
+  out_eigvecs.transpose_(0, 1);
```
I think it does: note that we are calling `transpose_()` twice: one in `eig_cuda` and one in `apply_eig`, to turn them back to row-order.
By doing so, we can pass the underlying `data_ptr<>` directly to `magmaEig<>` as `wr` and `wi`, thus avoiding a copy later.
This is similar to the old implementation of `THCTensor_(geev)`, which:
1. allocate `*wr` and `*wi` and fill them in column-order ([source](https://github.com/pytorch/pytorch/blob/master/aten/src/THC/generic/THCTensorMathMagma.cu#L131))
2. transpose later to turn the tensor into row-order ([source](https://github.com/pytorch/pytorch/blob/master/aten/src/THC/generic/THCTensorMathMagma.cu#L180))

I am assuming that `transpose_(0, 1)` does **not** copy any data, is this correct?

antocuni replied:
> Re the need for transpose: I think that transpose in unnecessary if you would use
> 
> ```
> out_eigvals.resize_({n, 2});
> ```
> 
> in `apply_eig`. Notice that `magma_Eig` uses only a `data_ptr` which will be the same with or without `transpose_`, hence you can create `out_eigvals` with the correct shape immediately.

I don't think so: the problem is that `magmaEig` writes the output values in column-ordered arrays. If I do what you suggest, I get a tensor of the correct shape, but with "transposed" values. I.e.:

```pycon
>>> x = torch.tensor([[1., 2, 3], [4, 5, 6], [7, 8, 9]], device='cuda')
>>> print(x.eig()) # my code
tensor([[ 1.6117e+01,  0.0000e+00],
         [-1.1168e+00,  0.0000e+00],
         [-1.2253e-07,  0.0000e+00]], device='cuda:0')

>>> print(x.eig()) # without the transpose
tensor([[ 1.6117e+01, -1.1168e+00],
        [-1.2253e-07,  0.0000e+00],
        [ 0.0000e+00,  0.0000e+00]], device='cuda:0'),
```

pearu replied:
so, it means that the strides of `x.eig()` needs to be reversed to get the values in the correct order.

antocuni commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1263,6 +1289,94 @@ std::tuple<Tensor, Tensor> _symeig_helper_cuda(const Tensor& self, bool eigenvec
+std::tuple<Tensor &,Tensor &> eig_cuda_out(Tensor & e, Tensor & v, const Tensor & self, bool eigenvectors) {
+  Tensor vals_tmp, vecs_tmp;
+  std::tie(vals_tmp, vecs_tmp) = eig_cuda(self, eigenvectors);
+  e.resize_as_(vals_tmp).copy_(vals_tmp);
+  v.resize_as_(vecs_tmp).copy_(vecs_tmp);
```
> In general, an important reason for using out argument is to avoid unnecessary memory copies.

I agree; what I did was to follow what it is done by the [existing `symEig`](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/BatchLinearAlgebra.cpp#L922), so I assumed it was "fine".

> IIUC, the pytorch convention of specifying out parameter is as follows [...]

good to know: I'll change the code to do that

t-vi replied:
The problem here (and the reason why it is symeig works how it does) is that vals and vecs might not be contiguous, in which case you cannot expect the backend to deal with them correctly.

antocuni replied:
wouldn't it be possible to check whether `out_vals` and `out_vecs` have the correct shape/format, and fall back to the copy approach only if the don't?

t-vi replied:
Possibly, yes, and it is nice to look at avoiding copies if possible. On the other hand, it isn't entirely clear that it is as crucial for n^3 algorithms as it is in the cheaper functions.
Maybe @vishwakftw has some insight to share from his work on `symeig`.

t-vi replied:
I must admit that altering the behavior of a single function to deviate from the PyTorch convention in favour of some Numpy condition did not occur to me as the ideal course of action for moving from TH to ATen.

pearu replied:
pytorch and numpy use different conventions in various places and while there exist efforts to unify these conventions, I agree that it makes sense to preserve existing conventions when porting to ATen.

antocuni replied:
so, if I understand it correctly:
- currently, it is supported to pass a non-contiguous tensor, and we expect `.eig` to "fix" it
- in the future, we want to deprecate this behavior and raise an error and/or warning if you pass the "wrong" out vector.

So I propose to already insert the check now:
- if the out vector is "correct", we will use it directly
- if it has not the correct shape/contiguity/whatever, we will use `copy` for now, and raise a warning and/or error in the future.

What do you think?

antocuni replied:
I removed the unnecessary copy in the `out=...` case, see commits ce66134 and 709510f. Now there is a `eig_cuda_helper` function which returns *CPU* tensors (since this is what is returned by `magmaEig` anyway), and `eig_cuda` and `eig_cuda_out` take care of copying the values to the right place.

Moreover, I am using `at::native::resize_output` to handle the output tensors: this way we get a warning in case the shape of the out tensor is not correct. In order to use it, I had to mark it as `C10_EXPORT` in `Resize.h`, because we need to link it from `libtorch_cpu.so` to `libtorch_cuda.so`.

If you agree with the changes, I'm going to mark this conversation as resolved :)

t-vi replied:
Hmm. Are we exporting at::native in general?

antocuni replied:
I don't know what is the official convention, but a quick `grep CAFFE2_API` inside `ATEN/native` shows that there are already exported functions in that namespace:
```
$ grep CAFFE2_API *.h
DispatchStub.h:struct CAFFE2_API DispatchStub;
DispatchStub.h:struct CAFFE2_API DispatchStub<rT (*)(Args...), T> {
DispatchStub.h:  extern CAFFE2_API struct name name
Resize.h:CAFFE2_API void resize_output(Tensor& output, IntArrayRef shape);
TensorIterator.h:struct CAFFE2_API OperandInfo {
TensorIterator.h:struct CAFFE2_API TensorIterator {
TensorIterator.h:class CAFFE2_API TensorIteratorConfig final {
TensorIterator.h:struct CAFFE2_API SplitUntil32Bit {
TensorIterator.h:  struct CAFFE2_API iterator {
TypeProperties.h:CAFFE2_API ResultTypeState update_result_type_state(const Tensor& tensor, const ResultTypeState& in_state);
TypeProperties.h:CAFFE2_API ScalarType result_type(const ResultTypeState& state);
TypeProperties.h:CAFFE2_API ScalarType result_type(TensorList tensors);
```

antocuni replied:
@t-vi can I mark this conversation as resolved, or do you want to add more?

antocuni commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1263,6 +1289,94 @@ std::tuple<Tensor, Tensor> _symeig_helper_cuda(const Tensor& self, bool eigenvec
+  });
+  singleCheckErrors(info, "eig_cuda");
+
+  if (eigenvectors) {
+      return std::tuple<Tensor, Tensor>(out_eigvals.to(self.device()), out_eigvecs.to(self.device()));
+  } else {
+      return std::tuple<Tensor, Tensor>(out_eigvals.to(self.device()), at::empty({0}, self.options()));
+  }
```
1. copying from GPU to CPU is [what the old code does](https://github.com/pytorch/pytorch/blob/master/aten/src/THC/generic/THCTensorMathMagma.cu#L175): shouldn't I just follow the same logic, and possibly add the optimization in another PR?

2. is this a correct optimization, anyway? I assumed that the output Tensor should be on the same device as the input one, isn't it the case?

t-vi replied:
Note that the MAGMA driver routines are hybrid algorithms and potentially have a somewhat complex interplay between CPU and GPU. Here the comment in the SymEig port might be likely to apply to Eig as well:
https://github.com/pytorch/pytorch/blob/9a5a7328660d4c205cb2ec2c3db341fb00ea8d3f/aten/src/ATen/native/cuda/BatchLinearAlgebra.cu#L1236-L1243

More generally: Should we expect the code to be very parallel between symeig and eig? To the point that we might share things with templates?

@pearu What is your take on what is the required device here?

pearu replied:
AFAIK, magma does not have `_gpu` variant of `geev` like `syev` has. Hence we cannot make `symeig` and `eig` code as alike as one would wish.

> * copying from GPU to CPU is [what the old code does](https://github.com/pytorch/pytorch/blob/master/aten/src/THC/generic/THCTensorMathMagma.cu#L175): shouldn't I just follow the same logic, and possibly add the optimization in another PR?

Let me copy from above: While the THC implementation is a good reference for porting, if there is a way to improve the existing code with respect to efficiency/simplicity, I think one should use the opportunity, IMHO.

> * is this a correct optimization, anyway? I assumed that the output Tensor should be on the same device as the input one, isn't it the case?

Assume that the input device is GPU. The answer depends on what users want to do with the output Tensor: (i) continue with the computations on GPU or (ii) access the output data from CPU. For the case (ii), copying the output to GPU will be unnecessary as for accessing the data, it must be copied back to CPU. For the case (i), the routines using the output from eig likely handle the copy of CPU data to GPU automatically anyway. Hence, keeping the output data where it was created seems more optimal to me than insisting that the input and output data must have the same locations, especially because in either way users do not need to change their code.

t-vi replied:
You are right about geev needing to move the input to CPU, but this:
> For the case (ii), copying the output to GPU will be unnecessary as for accessing the data, it must be copied back to CPU.

Are you very sure about that? You are advising people to introduce BC breaking changes here. I must admit I have doubts about it.

pearu replied:
@t-vi , thanks for pointing out the BC breakage issue.

mruberry replied:
Sorry to interject here, but have we considered reusing PyTorch's CPU dgeev (in TH) here? That is, have we actually compared the performance of GPU -> CPU -> TH dgeev -> GPU vs. this MAGMA-based approach? 

As for the out argument, by the way, PyTorch is moving to a model where we will not resize given outputs (unless they have no storage, in which case we interpret it as the user requesting a resize). However, resizing outputs or, in some cases, having to copy to them is currently deprecated, and it may more appropriate to throw a warning if that behavior was previously supported.

pearu replied:
yes, marking this conversation resolved.
Benchmarking `GPU -> CPU -> TH dgeev -> GPU vs. this MAGMA-based` would be valuable, imho.

antocuni committed:
fix test_scalar_check_cuda: previously, it raises IndexError instead of RuntimeError

antocuni committed:
Merge remote-tracking branch 'upstream/master' into antocuni/eig-aten

antocuni committed:
use assertEqual on data_ptr instead of assertIs, as discussed in https://github.com/pytorch/pytorch/pull/44105#discussion_r483572403

antocuni committed:
the old test_eig used to run only on CPU, thus the CUDA version lacked many tests. Split test_eig into 4 smaller tests, and @onlyCPU only two of them

antocuni committed:
fix the case eig(eigenvectors=False): this bug was uncovered thanks to the tests which were unskipped by d430455

antocuni committed:
improve eig_cuda_out by:

1. use at::native:resize_output to handle the out tensors, to emit a warning in case the shape is not correct

2. avoid an unecessary copy: now we have eig_cuda_helper which returns CPU tensors, and eig_cuda and eid_cuda_out can copy them to GPU tensors only once, at the end

antocuni committed:
move eig_cuda_helper up, so we can kill the forward declaration

antocuni committed:
write a test for what happens to out_evecs if eigenvectors=False; the test passed already on the CPU implementation, but failed on the CUDA one. Fix it

antocuni commented:
I think that the PR starts to be in a good shape, so I promoted it from "draft" to "ready to review". Hopefully all the CI tests should pass

pearu commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1263,6 +1290,123 @@ std::tuple<Tensor, Tensor> _symeig_helper_cuda(const Tensor& self, bool eigenvec
+ */
+static std::tuple<Tensor,Tensor> eig_cuda_helper(const Tensor & self, int64_t n, bool eigenvectors) {
+  // copy self to pinned CPU memory
+  auto self_working_copy = at::empty(self.sizes(),
+                                     at::TensorOptions(at::kCPU).dtype(self.dtype()).pinned_memory(true));
+
+  self_working_copy.transpose_(0, 1); // make it column-ordered, what magmaEig wants
+  self_working_copy.copy_(self);
```
The purpose of `transpose_` here is to fix the strides to make `self_working_copy` column-ordered tensor. As a side-effect, its data content of undefined values is changed as well. This is unnecessary.

Suggestion: use `auto self_working_copy = at::empty_strided(sizes, strides);` where `sizes == self.sizes()` and `strides=reversed(self.strides())` so that the inplace transpose can be eliminated.

antocuni replied:
good point, I wasn't aware of `at::empty_strided`. Fixed in 3ee78b0

pearu commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1263,6 +1290,123 @@ std::tuple<Tensor, Tensor> _symeig_helper_cuda(const Tensor& self, bool eigenvec
+  self_working_copy.copy_(self);
+
+  // tensors holding the results
+  auto out_eigvals = at::empty(0, self.options().device(at::kCPU), LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+  auto out_eigvecs = eigenvectors
+                     ? at::empty({n, n}, self.options().device(at::kCPU), LEGACY_CONTIGUOUS_MEMORY_FORMAT)
+                     : Tensor();
```
Same here. Using `at::empty_strided` with reversed strides one can eliminate related transpose operations.

antocuni replied:
fixed by 59ec066

pearu reviewed (dismissed):
The usage of `transpose_` on empty tensors involves unnecessary transpose of data while the intention is just to fix strides of the tensors. I suggest using `empty_strides` with reversed strides to ensure the column-ordered data storage so that all related transpose operations can be eliminated.

antocuni committed:
use CAFFE2_API instead of C10_EXPORT, I think this is the correct way to do it

antocuni committed:
remove the call to transpose_() by creating self_working_copy directly with the correct stride

antocuni committed:
use empty_strided also for out_eig{vals,vecs} to create them with the correct strides and be able to kill the calls to transpose_()

antocuni committed:
Merge remote-tracking branch 'upstream/master' into antocuni/eig-aten

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -5524,14 +5524,14 @@
   dispatch:
     CPU: legacy::cpu::_th_eig_out
-    CUDA: legacy::cuda::_th_eig_out
+    CUDA: eig_cuda_out
```
Are these custom CUDA dispatches necessary?

antocuni replied:
well, we have two different implementations, one for CPU and one for CUDA. Is there another way to dispatch to these?

mruberry commented on test/test_torch.py:
```diff
@@ -15936,10 +15936,10 @@ def transformation_cuda_fn(tensor, **kwargs):
+    @skipCUDAIfNoMagma
     @dtypes(torch.double)
-    def test_eig(self, device, dtype):
+    def test_eig_basic(self, device, dtype):
```
Let's move the eig tests to test_linalg.py

antocuni replied:
done in 17e3d511a5

mruberry commented on test/test_torch.py:
```diff
@@ -16000,6 +16010,37 @@ def test_eig(self, device, dtype):
+        evals, evecs = torch.eig(t, eigenvectors=True)
+        #
+        # check that the out= version computes the same values as the normal one
+        out_evals = torch.full_like(evals, fill_value=math.nan)
```
torch.empty_like?

antocuni replied:
done in 17e3d511a5

mruberry commented on test/test_torch.py:
```diff
@@ -16000,6 +16010,37 @@ def test_eig(self, device, dtype):
+        self.assertEqual(evecs, out_evecs)
+        #
+        # check what happens in the eigenvectors=False case
+        out_evals = torch.full_like(evals, fill_value=math.nan)
```
empty_like here, too

antocuni replied:
done in 17e3d511a5

mruberry commented on test/test_torch.py:
```diff
@@ -16000,6 +16010,37 @@ def test_eig(self, device, dtype):
+    @skipCUDAIfNoMagma
+    @skipCPUIfNoLapack
+    @dtypes(torch.double)
+    def test_eig_out(self, device, dtype):
```
This should also test that a warning is thrown when a non-empty out argument with the incorrect size throws a warning, and that an empty out argument is resized without a warning

antocuni replied:
Currently, the code does not do that: it simply resizes the output tensors (as it did in the original TH code).
I seem to understand that there is a plan to change the logic to forbid passing incorrectly-sized out tensors everywhere, but I expect that once this is implemented there will be a global helper function which you can just call on each out tensor and does the resizing/warning/error/whatever.
Isn't it better to wait to have this common logic instead of writing a custom version on each operator?

mruberry commented on test/test_torch.py:
```diff
@@ -15956,7 +15956,10 @@ def test_eig(self, device, dtype):
+    @onlyCPU
+    @skipCPUIfNoLapack
+    @dtypes(torch.double)
+    def test_eig_reuse(self, device, dtype):
```
Is there a test that (1) we can run in float precision, too, and (2) we can use a NumPy's eig as a reference? Maybe not by comparing the functions directly but, since the functions compute the same thing, validating the computations are close?

antocuni replied:
done by ca7f4991e5 (and 0f8148ab41 for an additional comment): I enabled `torch.float` for all eig tests (because, why not?)
I had to use an rtol!=0 for a few torch.float tests when comparing the results, though.

mruberry commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1281,6 +1308,118 @@ std::tuple<Tensor, Tensor> _symeig_helper_cuda(const Tensor& self, bool eigenvec
+  return std::tuple<Tensor&, Tensor&>(e, v);
+}
+
+std::tuple<Tensor,Tensor> eig_cuda(const Tensor & self, bool eigenvectors) {
```
Tensor& self

mruberry replied:
Can this function just allocate the result and call eig_cuda_out?

antocuni replied:
I think that the main reason why it didn't do it already is that I did a
straight port of the old TH code, and in TH in was done this way. But I agree
it makes sense, fixed by 451197f9f5.

The only small drawback is that we do the input check twice, in in eig_cuda
and one in eig_cuda_out. But we need to do it early in order to get the `n` to
allocate the results. It is probably avoidable by creating yet another helper
function, but I think that the performance penality is negligible here.

mruberry commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1281,6 +1308,118 @@ std::tuple<Tensor, Tensor> _symeig_helper_cuda(const Tensor& self, bool eigenvec
+}
+
+std::tuple<Tensor,Tensor> eig_cuda(const Tensor & self, bool eigenvectors) {
+  TORCH_CHECK(self.dim() == 2, "A should be 2 dimensional");
```
"Expected a two-dimensional input but got X dimensions!" (where X is the number of dimensions that self has)

antocuni replied:
fixed by 89ca976835

mruberry commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1281,6 +1308,118 @@ std::tuple<Tensor, Tensor> _symeig_helper_cuda(const Tensor& self, bool eigenvec
+  if (eigenvectors) {
+      return std::tuple<Tensor, Tensor>(cpu_vals.to(self.device()), cpu_vecs.to(self.device()));
+  } else {
+      return std::tuple<Tensor, Tensor>(cpu_vals.to(self.device()), at::empty({0}, self.options()));
```
Technically I think the empty tensor should use cpu_vecs options but with the device of self, right?

antocuni replied:
true! Although this code has been deleted now :)

mruberry commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1281,6 +1308,118 @@ std::tuple<Tensor, Tensor> _symeig_helper_cuda(const Tensor& self, bool eigenvec
+  return std::tuple<Tensor, Tensor>(out_eigvals, out_eigvecs);
+}
+
+std::tuple<Tensor &,Tensor &> eig_cuda_out(Tensor & e, Tensor & v, const Tensor & self, bool eigenvectors) {
```
Tensor&

antocuni replied:
b2a16c1e4f

mruberry commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1281,6 +1308,118 @@ std::tuple<Tensor, Tensor> _symeig_helper_cuda(const Tensor& self, bool eigenvec
+  Tensor cpu_vals, cpu_vecs;
+  std::tie(cpu_vals, cpu_vecs) = eig_cuda_helper(self, n, eigenvectors);
+  e.copy_(cpu_vals);
+  if (eigenvectors)
```
Every if needs curly braces:

```
if (...) {
  ...
}
```

antocuni replied:
b2a16c1e4f

mruberry commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1281,6 +1308,118 @@ std::tuple<Tensor, Tensor> _symeig_helper_cuda(const Tensor& self, bool eigenvec
+  int64_t n = self.size(-1);
+
+  at::native::resize_output(e, {n, 2});
+  if (eigenvectors)
```
Curly braces

antocuni replied:
b2a16c1e4f

mruberry commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1281,6 +1308,118 @@ std::tuple<Tensor, Tensor> _symeig_helper_cuda(const Tensor& self, bool eigenvec
+
+  return std::tuple<Tensor, Tensor>(out_eigvals, out_eigvecs);
+}
+
```
Add a comment describing the approach

antocuni replied:
tentatively addressed by bf8d35f9f3

mruberry commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1281,6 +1308,118 @@ std::tuple<Tensor, Tensor> _symeig_helper_cuda(const Tensor& self, bool eigenvec
+/*
+ * Internal helper; like eig_cuda but:
+ *   1. assume that self is a square matrix of side "n"
+ *   2. return CPU tensors, which will be copied to GPU memory by the caller
```
This needs to explain why CPU tensors are returned

antocuni replied:
bf8d35f9f3

mruberry commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1281,6 +1308,118 @@ std::tuple<Tensor, Tensor> _symeig_helper_cuda(const Tensor& self, bool eigenvec
+static void apply_eig(const Tensor& self, bool eigenvectors, Tensor& out_eigvals, Tensor& out_eigvecs,
+                      int64_t *info_ptr) {
+#ifndef USE_MAGMA
+AT_ERROR("symeig: MAGMA library not found in "
```
AT_ERROR is deprecated and this refers to symeig, not eig. 

```
TORCH_CHECK(false, "...");
```

I know the form of this error message is inherited, but it's terrible so let's update it. What about:

"Calling torch.eig on a CUDA tensor requires compiling PyTorch with MAGMA. Either transfer the tensor to the CPU before calling torch.eig or recompile with MAGMA."?

antocuni replied:
fixed by 89ca976835. Note that all the other functions in the same file are
still using AT_ERROR and the old error message

mruberry commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1281,6 +1308,118 @@ std::tuple<Tensor, Tensor> _symeig_helper_cuda(const Tensor& self, bool eigenvec
 }
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ eig ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
```
Brief description of what this function does - and why it takes a CPU tensor.  Should this assert that self is on the CPU? 

Also, the name "apply_eig" is very generic. Is there a better, more CUDA-specific name for this function? "apply_eig_using_cuda"?

antocuni replied:
>  why it takes a CPU tensor. Should this assert that self is on the CPU?

done by 6b5f0ce055. Is there a better way to do an assert, or `TORCH_CHECK` is fine?

> Also, the name "apply_eig" is very generic. Is there a better, more CUDA-specific name for this function? "apply_eig_using_cuda"?

maybe, but note that this is the naming convention of the whole file: every XXX operator has an `apply_XXX` function, so I think it's better to stick with it

mruberry replied:
TORCH_CHECK is correct for assertions that are a result of user behavior, TORCH_INTERNAL_ASSERT is for assertions that are due to internal logical errors.

mruberry commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1281,6 +1308,118 @@ std::tuple<Tensor, Tensor> _symeig_helper_cuda(const Tensor& self, bool eigenvec
+  // optimization: if self is empty, we can immediately allocate&return empty
+  // GPU tensors, instead of getting empty CPU tensors from eig_cuda_helper
+  // and copying them to GPU
+  if (self.numel() == 0) {
```
It's interesting that this short-circuit doesn't appear in the out variant. Having this function call the out variant should prevent these discrepancies.

antocuni replied:
true, fixed by 451197f9f5.

mruberry reviewed (commented):
Hey @antocuni!

Overall this look pretty good. I made a note about moving the tests and expanding their coverage, as well as a few notes about reusing the out variant to avoid code duplication and providing comments about the CUDA->CPU->CUDA data movement. 

We should also take this opportunity to update the torch.eig documentation. First, let's kill the "return type" section. Second, let's add a couple examples (one with the default values and another with eigenvectors=True). Third, let's add a warning that when called with a CUDA input this, unlike most PyTorch functions, will synchronize the input's device with the CPU. We try very hard to avoid that and should be clear about when it's happening.

I'm also not in love with torch.eig's description of eigenvalues and eigenvectors, but maybe we'll deal with those when performing the CPU port.

antocuni committed:
move test_eig_* from test_torch to test_linalg and use empty_like
instead of full_like

antocuni committed:
- run all eig tests also with dtype=torch.float; we need to increase a bit the rtol to make a couple of them passing, though

- test eig against numpy

antocuni committed:
cosmetics

antocuni committed:
add a comment

antocuni committed:
implement eig_cuda in terms of eig_cuda_out

antocuni committed:
improve error messages

antocuni committed:
try to improve the comments

antocuni committed:
add a sanity check that we receive a CPU tensor here

antocuni committed:
try to improve the docs

antocuni commented:
Hello @mruberry, sorry for the delay in the response. I didn't realize that this PR was higher-prio than the other one I was working on.

I think I addressed most of your suggestions, apart the one about emitting a warning when resizing the out tensors (see my inline response).

> We should also take this opportunity to update the torch.eig documentation. First, let's kill the "return type" section. Second, let's add a couple examples (one with the default values and another with eigenvectors=True). Third, let's add a warning that when called with a CUDA input this, unlike most PyTorch functions, will synchronize the input's device with the CPU. We try very hard to avoid that and should be clear about when it's happening.

I tried to integrate these suggestions in aabdd4f0f6. For the examples, I have used the very same as numpy

antocuni committed:
Merge upstream/master into antocuni/eig-aten

antocuni committed:
try to fix flake8

mruberry commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1587,6 +1614,125 @@ std::tuple<Tensor, Tensor> _symeig_helper_cuda(const Tensor& self, bool eigenvec
+TORCH_CHECK(false, "Calling torch.eig on a CUDA tensor requires compiling PyTorch with MAGMA. "
+                   "Either transfer the tensor to the CPU before calling torch.eig or recompile with MAGMA.");
+#else
+  TORCH_CHECK(self.device() == at::kCPU, "Internal error: apply_eig needs a CPU tensor");
```
TORCH_INTERNAL_ASSERT if this error really does catch an internal logic failure vs. something the user could be expected to trigger.

antocuni replied:
6c0e45cf06

mruberry commented on test/test_linalg.py:
```diff
@@ -674,6 +675,122 @@ def test_norm_fastpaths(self, device):
 
+    @skipCPUIfNoLapack
+    @skipCUDAIfNoMagma
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
```
The test suite now requires NumPy so we no longer need checks like this

antocuni replied:
e71ac6c6cc

mruberry commented on test/test_linalg.py:
```diff
@@ -674,6 +675,122 @@ def test_norm_fastpaths(self, device):
+    @skipCPUIfNoLapack
+    @skipCUDAIfNoMagma
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
+    @dtypes(torch.double, torch.float)
```
Might as well just do torch.float since the logic is the same for both dtypes?

mruberry commented on test/test_linalg.py:
```diff
@@ -674,6 +675,122 @@ def test_norm_fastpaths(self, device):
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
+    @dtypes(torch.double, torch.float)
+    def test_eig_basic(self, device, dtype):
+        a = torch.Tensor(((1.96, 0.00, 0.00, 0.00, 0.00),
```
`torch.tensor` instead of `Tensor`, instead of using `to()` you can pass the `dtype` and `device` kwargs to the `tensor()` constructor directly

antocuni replied:
uh, you are correct. This is a test which was already there, I don't know why they wrote it like this in the first place.
Fixed by 63ca2b160d

mruberry commented on test/test_linalg.py:
```diff
@@ -674,6 +675,122 @@ def test_norm_fastpaths(self, device):
+    @skipCUDAIfNoMagma
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
+    @dtypes(torch.double, torch.float)
+    def test_eig_basic(self, device, dtype):
```
Can this test and the next two be simplified by enumerating test cases and just comparing with NumPy? 

The test cases could include discontiguous tensors. 

I think the out variant of eig could also be tested at the same time. Curious to hear your thoughts on this.

mruberry reviewed (commented):
Hey @antocuni! Changes look good. Made a few additional cleanup comments.

antocuni committed:
use TORCH_INTERNAL_ASSERT

antocuni committed:
Merge remote-tracking branch 'upstream/master' into antocuni/eig-aten

antocuni committed:
no longer needed

antocuni committed:
use a more direct way to create the tensor

antocuni commented:
Thank you. I fixed some of your remarks, let me comment here on the others.

> Might as well just do torch.float since the logic is the same for both dtypes?

I'm not sure I agree, for two reasons:
1. the logic is the same now, but potentially in the future someone might refactor the code in a way which breaks only floats or only doubles, and the tests woult not catch it

2. the logic is essentially the same for 99.9%, but we also want to test the other 0.01%. In particular, we want to ensure that we call the correct lapack routine (especially considering that it's very easy to do a mistake, since lapack's names are quite obscure and short).


> Can this test and the next two be simplified by enumerating test cases and just comparing with NumPy?
> The test cases could include discontiguous tensors.
> I think the out variant of eig could also be tested at the same time. Curious to hear your thoughts on this.

so, or master there is a single giant test which tests all the different cases, with comments inside to split the various "basic", "reuse", "non-contiguous", etc. cases. I preferred to split it into multiple tests to make each test easier to read (since it's more self-contained and you don't have to look 20 lines up to see what happened) and the development more manageable, since it's easier to make tests passing one by one that everything at once.

As for just comparing to numpy: maybe, but again I find these kind of tests very hard to debug. When they fail, all you see is that two tensors with random numbers are different and you have no idea why. So what you do is to find the place where you enumerate all the cases and comment out them one by one until you find the culprit, and then you are basically back to have your single&small test.

At the end of the day is a matter of style preference: personally I have a preference for tests which are shorter&smaller and that contains the least amount possible of extra logic, but feel free to disagree and decide otherwise if you prefer.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -2698,6 +2701,32 @@ def merge_dicts(*dicts):
+
+Example::
+
+    Trivial example with a diagonal matrix. By default, only eigenvalues are computed:
```
Cool example

antocuni replied:
I ported it straight from numpy docs

mruberry commented on torch/_torch_docs.py:
```diff
@@ -2674,6 +2674,9 @@ def merge_dicts(*dicts):
     Since eigenvalues and eigenvectors might be complex, backward pass is supported only
     for :func:`torch.symeig`
 
+    When :attr:`input` is on CUDA, :func:`torch.eig() <torch.eig>` causes
```
This is a great note. 

cc @heitorschueroff who's reviewing operations using MAGMA for host-device sync.

mruberry reviewed (approved):
Hey @antocuni, the updates look great and I appreciate your thinking. 

Keeping the tests as they are for now sounds fine. We'll have the opportunity to make additional test improvements when we holistically review test_linalg.py and update it to use OpInfos, too.

cc @kurtamohler, we should be OK to move forward with https://github.com/pytorch/pytorch/pull/43215 now, too.

mruberry commented:
Unfortunately test_linalg.py is a popular place these days. Would you please rebase this, @antocuni?

antocuni committed:
Merge remote-tracking branch 'upstream/master' into antocuni/eig-aten

antocuni commented:
> Unfortunately test_linalg.py is a popular place these days. Would you please rebase this, @antocuni?

done!

mruberry commented:
Bad news: this had to be reverted.

This appears to have broken two builds:

ci/circleci: pytorch_linux_xenial_cuda10_2_cudnn7_py3_nogpu_NO_AVX_test
and
ci/circleci: pytorch_linux_xenial_cuda10_2_cudnn7_py3_nogpu_NO_AVX2_test

These are CUDA-capable builds of PyTorch on machines that don't actually GPUs and where AVX or AVX2 is disabled or unavailable.

test failures:
```
test_eig_non_contiguous_cpu_float32 - TestLinalgCPU
test_linalg.py

Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 272, in instantiated_test
    result = test_fn(self, *args)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 631, in only_fn
    return fn(slf, device, *args, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 525, in dep_fn
    return fn(slf, device, *args, **kwargs)
  File "test_linalg.py", line 1068, in test_eig_non_contiguous
    self.assertEqual(X, Xhat, atol=1e-8, rtol=rtol, msg='VeV\' wrong')
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1112, in assertEqual
    self.assertTrue(result, msg=msg)
AssertionError: False is not true : VeV' wrong
		
test_eig_reuse_cpu_float32 - TestLinalgCPU
test_linalg.py

Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 272, in instantiated_test
    result = test_fn(self, *args)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 631, in only_fn
    return fn(slf, device, *args, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 525, in dep_fn
    return fn(slf, device, *args, **kwargs)
  File "test_linalg.py", line 1047, in test_eig_reuse
    self.assertEqual(X, Xhat, atol=1e-8, rtol=rtol, msg='VeV\' wrong')
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1112, in assertEqual
    self.assertTrue(result, msg=msg)
AssertionError: False is not true : VeV' wrong
```

Does anything here make sense, @antocuni? 

Unfortunately I don't think we run these builds in the github CI, which is why we didn't see any signs they would break before merging the PR.

mruberry commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1738,6 +1765,125 @@ std::tuple<Tensor, Tensor> _symeig_helper_cuda(const Tensor& self, bool eigenvec
+}
+
+std::tuple<Tensor&, Tensor&> eig_cuda_out(Tensor& e, Tensor& v, const Tensor& self, bool eigenvectors) {
+  TORCH_CHECK(self.dim() == 2, "Expected a two-dimensional input but got ", self.dim(), " dimensions");
```
@ngimel also mentioned that this should check that out is the same dtype, too

since this has been reverted we may as well fix that

antocuni replied:
fixed by 1e40ca4f80

antocuni committed:
Merge remote-tracking branch 'upstream/master' into antocuni/eig-aten

antocuni committed:
always include the debug_msg in the assert message when a tensor comparison fails. This makes it much easier to debug failures on CI jobs

antocuni committed:
check that the out dtypes match

antocuni commented:
>This appears to have broken two builds:
> ci/circleci: pytorch_linux_xenial_cuda10_2_cudnn7_py3_nogpu_NO_AVX_test
> and
> ci/circleci: pytorch_linux_xenial_cuda10_2_cudnn7_py3_nogpu_NO_AVX2_test

I **think** it's simply a problem loosing the tolerance for the `float32` case. Some considerations:

1. the two failing tests exist already on master (although with a different name, this branch split the old `test_eig` into multiple ones) but on master they are tested only with `dtype=double`. This makes me thinking that probably they would fail on master as well

2. the tests fail only for `float32` and I have already had to loosen the tolerance a bit for that case

3. the tests fail only on some peculiar hardware/environment. I suppose that what happens is that since the compiled code uses different instructions (e.g. SSE vs AVX) the result is slightly different, enough to surpass the tolerance.

I have tried hard to reproduce the problem on my system, trying to build pytorch with different CFLAGS and settings, but I didn't manage: the tests always fail for me. So I committed 23a767d235 which tries to improve the error message that we get: hopefully with it we just need to tweak the tolerance or if there is a more serious problem.

If I understand correctly, I don't have any way to trigger a build of the relevant jobs it by myself, do I? @mruberry could you please trigger it?

kurtamohler commented:
> If I understand correctly, I don't have any way to trigger a build of the relevant jobs it by myself, do I?

Does this help @antocuni ? https://github.com/Quansight/pytorch/wiki/Reproducing-Build-Failures-directly-on-the-CI-machines

antocuni commented:
> > If I understand correctly, I don't have any way to trigger a build of the relevant jobs it by myself, do I?
> 
> Does this help @antocuni ? https://github.com/Quansight/pytorch/wiki/Reproducing-Build-Failures-directly-on-the-CI-machines

I fear not, because IIUC it's a CI failure which happens on facebook's private CI, not the public one.

ngimel commented:
The failing builds are on public CI. Push your change to a branch named ci-all/XXXX (on pytorch/pytorch) and open a PR from it, and all jobs, not just PR-only branches, will trigger.

antocuni commented:
> The failing builds are on public CI. Push your change to a branch named ci-all/XXXX (on pytorch/pytorch) and open a PR from it, and all jobs, not just PR-only branches, will trigger.

wonderful, I didn't know. I opened the new PR #48402

antocuni committed:
increase a bit the float32 tolerance for these two tests

antocuni committed:
don't check the out_v.dtype if eigenvectors==False, and add a test

antocuni committed:
relax the error message regexp so that it matches also the errors you get in the CPU case

antocuni committed:
use the correct atol also here

mruberry commented:
Thanks for jumping through all these hoops, @antocuni!

It's interesting that all the tests are now running on this PR, too. I wonder why that is?

antocuni commented:
> It's interesting that all the tests are now running on this PR, too. I wonder why that is?

interesting, I didn't notice that. If you try to click on the individual jobs, you see different things:

- jobs such as [docker-pytorch-linux-bionic-cuda10.2-cudnn7-py3.8-gcc9](https://app.circleci.com/pipelines/github/pytorch/pytorch/242694/workflows/069633cf-4bd4-4fff-a9b0-6c564bec5c78/jobs/9148206) say:
    * branch: `pull/44105`
    * PR/Commit: `#44105 / c3e2637`

- jobs such as [pytorch_linux_xenial_cuda10_2_cudnn7_py3_nogpu_NO_AVX2_test](https://app.circleci.com/pipelines/github/pytorch/pytorch/242673/workflows/2291d126-013a-4595-8abd-6a6870e7283b/jobs/9146255) say:
    * branch: `ci-all/antocuni-eig-aten-cuda`
    * PR/Commit: `#48402 / c3e2637`

So my guess is that the extra tests were actually run in the **other** PR, but for some reason (possibly because it's the very same commit) they are shown also here
